oauth2client/django_util/__init__.py

```diff
@@ -0,0 +1,284 @@
+    """Initializes Django OAuth2 Helper Settings"""
+    def __init__(self, settings_instance):
+        self.scopes = getattr(settings_instance, 'GOOGLE_OAUTH2_SCOPES',
+                              GOOGLE_OAUTH2_DEFAULT_SCOPES)
+        self.request_prefix = getattr(settings_instance, 'GOOGLE_OAUTH2_REQUEST_ATTRIBUTE',
+                                      GOOGLE_OAUTH2_REQUEST_ATTRIBUTE)
+        self.client_id, self.client_secret = \
+            _get_oauth2_client_id_and_secret(settings_instance)
+
+        if 'django.contrib.sessions.middleware.SessionMiddleware' not in \
+                settings_instance.MIDDLEWARE_CLASSES:
+            raise ImproperlyConfigured(
+                "The Google OAuth2 Helper requires session middleware to "
+                "be installed. Edit your MIDDLEWARE_CLASSES setting"
+                " to include 'django.contrib.sessions.middleware."\
+                "SessionMiddleware'.")
+
+oauth2_settings = OAuth2Settings(settings)
+
+
+class UserOAuth2(object):
+    """ Class to create oauth2 objects on Django request objects containing
+    credentials and helper methods.
+    """
+
+    def __init__(self, request, scopes=None, return_url=None):
+        """ Initialize the Oauth2 Object
+        :param request: Django request object
+        :param scopes: Scopes desired for this OAuth2 flow
+        :param return_url: URL to return to after authorization is complete
+        :return:
+        """
+        self.request = request
+        self.return_url = return_url or request.get_full_path()
+        self.scopes = set(oauth2_settings.scopes)
+        if scopes:
+            self.scopes |= set(scopes)
+
+        # make sure previously requested custom scopes are maintained
+        # in future authorizations
+        credentials = get_storage(self.request).get()
+        if credentials:
+            self.scopes |= credentials.scopes
+
+    def get_authorize_redirect(self):
+        """Creates a URl to start the OAuth2 authorization flow"""
+        get_params = {
+            'return_url': self.return_url,
+            'scopes': self.scopes
+        }
+
+        return _redirect_with_params('google_oauth:authorize',
+                                     **get_params)
+
+    def has_credentials(self):
+        """Returns True if there are valid credentials for the current user and required scopes."""
+        return self.credentials and not self.credentials.invalid \
+            and self.credentials.has_scopes(self.scopes)
+
+    @property
+    def credentials(self):
+        """Gets the authorized credentials for this flow, if they exist"""
+        return get_storage(self.request).get()
+
+    @property
+    def http(self):
+        """Helper method to create an HTTP client authorized with OAuth2 credentials"""
+        if self.has_credentials():
+            return self.credentials.authorize(httplib2.Http())
+        return None
+
+
+def _redirect_with_params(url_name, *args, **kwargs):
```

Comment:
Please place this function definition above (earlier in the file text) all of its use-sites.